projects/vdk-plugins/vdk-data-sources/README.md

```diff
@@ -0,0 +1,91 @@
+
+The data-sources project is a plugin for the Versatile Data Kit (VDK). It aims to simplify data ingestion from multiple sources by offering a single, unified API. Whether you're dealing with databases, REST APIs, or other forms of data, this project allows you to manage them all in a consistent manner. This is crucial for building scalable and maintainable data pipelines.
+
+
+## Usage
+
+```
+pip install vdk-data-sources
+```
+
+### Concepts
+
+
+#### Data Source
+A Data Source is a central component responsible for establishing and managing a connection to a specific set of data. It interacts with a given configuration and maintains a stateful relationship with the data it accesses. This stateful relationship can include information such as authentication tokens, data markers, or any other form of metadata that helps manage the data connection. The Data Source exposes various data streams through which data can be read.
+
+#### Data Source Stream
+A Data Source Stream is an abstraction over a subset of data in the Data Source. It can be thought of as a channel through which data flows. Each Data Source Stream has a unique name to identify it and includes methods to read data from the stream. For example for Database based data source , each table could be a separate stream. Streams can be ingested in parallel potentially.
+
+Reading from the stream yields a sequence of Data Source Payloads
+
+#### Data Source Payload
+The Data Source Payload is a data structure that encapsulates the actual data along with its metadata. The payload consists of four main parts:
+
+Data: containing the core data that needs to be ingested (e.g in database the table content)
+Metadata: A dictionary containing additional contextual information about the data (for example timestamps, environment specific metadata, etc.)
+State: Contains the state of the data soruce stream as of this payload. For example in case of incremental ingestion from a database table it would contain the value of a incremental key columns (le.g updated_time column in teh table) which can be used to restart/continue the ingestion later.
+
+
+### Configuration
+
+(`vdk config-help` is useful command to browse all config options of your installation of vdk)
+
+### Example
+
+To build your own data source you can use [this data source](./src/vdk/plugin/data_sources/auto_generated.py) as an example or reference
+
+To register the source use [vdk_data_sources_register hook](./src/vdk/plugin/data_sources/hook_spec.py)
+
+Then you can use it in a data job like this:
+
+```python
+def run(job_input: IJobInput):
+    source = SourceDefinition(id="auto", name="auto-generated-data", config={})
+    destination = DestinationDefinition(id="auto-dest", method="memory")
+
+    with DataFlowInput(job_input) as flow_input:
+        flow_input.start(DataFlowMappingDefinition(source, destination))
+```
+
+or in config.toml file
```

Comment:
I've been considering replacing ini with TOML because 
- it supported nested structures of data 
- it supports arrays 
- it supports data types !! 
- it's somewhat similar in ini in syntax - existing ini files can be parsed by TOML parser so a change can be pretty backward compatible.

It would not have been feasible/easy to use ini format the above data flow structure. So I decided to use TOML now as an experiment to see if it's going to work for users. 

We need to move away from ini due to above reasons and we've had users who have requested to support more "modern" format. 
The other alternative is Yaml. But yaml is pretty ugly for highly nested configurations an

Reply:
I generally agree with everything you said, however I would be great to have a more structured approach, e.g. use TOML everywhere, rather than use it here and use something else in a different place.